lib/spack/spack/database.py

```diff
@@ -1561,9 +1568,26 @@ def _query(
                 if not (start_date < inst_date < end_date):
                     continue
 
-            if query_spec is any or rec.spec.satisfies(query_spec):
+            if query_spec is any:
+                results.append(rec.spec)
+                continue
+
+            # save non-name matches for later
+            if query_spec.name and rec.spec.name != query_spec.name:
+                deferred.append(rec.spec)
+                continue
+
+            if rec.spec.satisfies(query_spec):
                 results.append(rec.spec)
```

Comment:
You can stop appending to `deferred` if `results` is non-empty, and specs in the db always have a name, so no need to check if `query_spec.name` is truthy every iteration.

```suggestion
            # only call satisfies on equal name to avoid expensive virtual check
            if rec.spec.name == query_spec.name
                if rec.spec.satisfies(query_spec):
                    results.append(rec.spec)
            elif not results:
                deferred.append(rec.spec)
```

Reply:
Tried this but it breaks tests -- reverting... I don't think the optimizations with `deferred` or with the name check are all that important... so would rather just merge as is.

Reply:
Let me know if you want me to reinvestigate this.

Reply:
See below